File: PlayPcmWin/WWDirectComputeCS/WWUpsampleCpu.cs
using System;
using System.Threading.Tasks;

namespace WWDirectComputeCS
{
    public class WWUpsampleCpu
    {
        private static void
        PrepareResamplePosArray(
                int sampleTotalFrom,
                int sampleRateFrom,
                int sampleRateTo,
                int sampleTotalTo,
                int [] resamplePosArray,
                double [] fractionArrayD) {
            for (int i = 0; i < sampleTotalTo; ++i) {
                double resamplePos = (double)i * sampleRateFrom / sampleRateTo;
                /* -0.5 <= fraction<+0.5になるようにresamplePosを選ぶ。
                 * 最後のほうで範囲外を指さないようにする。
                 */
                int resamplePosI = (int)(resamplePos + 0.5);
                if (resamplePosI < 0) {
                    resamplePosI = 0;
                }
                if (sampleTotalFrom <= resamplePosI) {
                    resamplePosI = sampleTotalFrom - 1;
                }
                double fraction = resamplePos - resamplePosI;

                resamplePosArray[i] = resamplePosI;
                fractionArrayD[i] = fraction;
            }
        }

        private static void
        PrepareSinPreComputeArray(
                double []fractionArray,
                int sampleTotalTo,
                float []sinPreComputeArray)
        {
            for (int i=0; i<sampleTotalTo; ++i) {
                sinPreComputeArray[i] = (float)Math.Sin(-Math.PI * fractionArray[i]);
            }
        }

        private int m_convolutionN;
        private int m_sampleTotalFrom;
        private int m_sampleRateFrom;
        private int m_sampleRateTo;
        private int m_sampleTotalTo;
        private int [] m_resamplePosArray;
        private double [] m_fractionArray;
        private float [] m_sampleFrom;
        private float [] m_sinPreComputeArray;

        public int Setup(
                int convolutionN,
                float [] sampleFrom,
                int sampleTotalFrom,
                int sampleRateFrom,
                int sampleRateTo,
                int sampleTotalTo,
                int [] resamplePosArray,
                double [] fractionArrayD)
        {
            int hr = 0;

            System.Diagnostics.Debug.Assert(0 < convolutionN);
            System.Diagnostics.Debug.Assert(sampleFrom != null);
            System.Diagnostics.Debug.Assert(0 < sampleTotalFrom);
            System.Diagnostics.Debug.Assert(sampleRateFrom <= sampleRateTo);
            System.Diagnostics.Debug.Assert(0 < sampleTotalTo);

            m_convolutionN    = convolutionN;
            m_sampleTotalFrom = sampleTotalFrom;
            m_sampleRateFrom  = sampleRateFrom;
            m_sampleRateTo    = sampleRateTo;
            m_sampleTotalTo   = sampleTotalTo;

            m_resamplePosArray = resamplePosArray;

            m_fractionArray = fractionArrayD;

            m_sampleFrom      = sampleFrom;

            m_sinPreComputeArray = new float[sampleTotalTo];

            PrepareSinPreComputeArray(m_fractionArray, sampleTotalTo, m_sinPreComputeArray);

            return hr;
        }

        private static double
        SincD(double sinx, double x)
        {
            if (-2.2204460492503131e-016 < x && x < 2.2204460492503131e-016) {
                return 1.0;
            } else {
                return sinx / x;
            }
        }

        // without resamplePosArray
        public int Setup(
                int convolutionN,
                float [] sampleFrom,
                int sampleTotalFrom,
                int sampleRateFrom,
                int sampleRateTo,
                int sampleTotalTo)
        {
            int hr = 0;

            System.Diagnostics.Debug.Assert(0 < convolutionN);
            System.Diagnostics.Debug.Assert(null != sampleFrom);
            System.Diagnostics.Debug.Assert(0 < sampleTotalFrom);
            System.Diagnostics.Debug.Assert(sampleRateFrom <= sampleRateTo);
            System.Diagnostics.Debug.Assert(0 < sampleTotalTo);

            // 多少無駄だが…
            int []    resamplePosArray = new int[sampleTotalTo];
            double [] fractionArrayD   = new double[sampleTotalTo];

            PrepareResamplePosArray(
                sampleTotalFrom, sampleRateFrom, sampleRateTo, sampleTotalTo,
                resamplePosArray, fractionArrayD);

            hr = Setup(
                  convolutionN,
                  sampleFrom,
                  sampleTotalFrom,
                  sampleRateFrom,
                  sampleRateTo,
                  sampleTotalTo,
                  resamplePosArray,
                  fractionArrayD);

            fractionArrayD   = null;
            resamplePosArray = null;

            return hr;
        }

        public int Do(
                int startPos,
                int count,
                float [] output)
        {
            int hr = 0;

            Parallel.For(startPos, startPos + count, delegate(int toPos)
            {
                int fromPos = m_resamplePosArray[toPos];
                double fraction = m_fractionArray[toPos];
                double sinPreCompute = m_sinPreComputeArray[toPos];

                double v = 0.0;

                for (int convOffs = -m_convolutionN; convOffs < m_convolutionN; ++convOffs) {
                    int pos = convOffs + fromPos;
                    if (0 <= pos && pos < m_sampleTotalFrom) {
                        double x = Math.PI * (convOffs - fraction);

                        double sinX = sinPreCompute;
                        if (0 != (convOffs & 1)) {
                            sinX *= -1.0;
                        }

                        double sinc = SincD(sinX, x);

                        v += m_sampleFrom[pos] * sinc;
                    }
                }
                // output[0]～output[count-1]に書き込む。
                output[toPos - startPos] = (float)v;
            });

            return hr;
        }

        public void Unsetup()
        {
            m_sinPreComputeArray = null;
            m_fractionArray = null;
            m_resamplePosArray = null;
            m_sampleFrom = null;
        }

    }
}

File: PlayPcmWin/WWDirectComputeTest/WWUpsampleCpu.cpp
// 日本語。

#include "WWUpsampleCpu.h"
#include "WWUtil.h"
#include <assert.h>
#include <math.h>
#include <stdio.h>



HRESULT
WWUpsampleCpu(
        int convolutionN,
        float * sampleData,
        int sampleTotalFrom,
        int sampleRateFrom,
        int sampleRateTo,
        float * outputTo,
        int sampleTotalTo)
{
    HRESULT hr = S_OK;

    assert(sampleRateFrom <= sampleRateTo);

    unsigned int * resamplePosArray = new unsigned int[sampleTotalTo];
    assert(resamplePosArray);

    float * fractionArray = new float[sampleTotalTo];
    assert(fractionArray);

    double *sinPreComputeArray = new double[sampleTotalTo];
    assert(sinPreComputeArray);

    for (int i=0; i<sampleTotalTo; ++i) {
        double resamplePos = (double)i * sampleRateFrom / sampleRateTo;
#if 1
        /* -0.5 <= fraction<+0.5になるようにresamplePosを選ぶ。
         * 最後のほうで範囲外を指さないようにする。
         */
        int resamplePosI = (int)(resamplePos+0.5);
        if (sampleTotalFrom <= resamplePosI) {
            resamplePosI = sampleTotalFrom -1;
        }
#else
        /* 0<=fraction<1になるにresamplePosIを選ぶ。
         * これは1に近い値が頻出するのでよくない。
         */
        int resamplePosI = (int)(resamplePos+0.5);
        assert(resamplePosI < sampleTotalFrom);
#endif
        double fraction = resamplePos - resamplePosI;

        resamplePosArray[i]   = resamplePosI;
        fractionArray[i]      = (float)fraction;
        sinPreComputeArray[i] = sin(-PI_D * fraction);
    }

    /*
    for (int i=0; i<sampleTotalTo; ++i) {
        printf("i=%6d rPos=%6d fraction=%+f\n",
            i, resamplePosArray[i], fractionArray[i]);
    }
    */
    //printf("resamplePos created\n");

    for (int toPos=0; toPos<sampleTotalTo; ++toPos) {
        int    fromPos  = resamplePosArray[toPos];
        double fraction = fractionArray[toPos];
        double sinPreCompute = sinPreComputeArray[toPos];

        double v = 0.0;

        for (int convOffs=-convolutionN; convOffs < convolutionN; ++convOffs) {
            int pos = convOffs + fromPos;
            if (0 <= pos && pos < sampleTotalFrom) {
                double x = PI_D * (convOffs - fraction);
                
                double sinX = sinPreCompute;
                if (convOffs & 1) {
                    sinX *= -1.0;
                }

#if 1
                // 合っていた。
                assert(fabs(sinX - sin(x)) < 0.000001);
#endif

                double sinc =  SincD(sinX, x);

                /*
                if (pos == 0) {
                    printf("toPos=%d pos=%d x=%f sinX=%f",
                        toPos, pos, x, sinX);
                    printf("\n");
                }
                */

                v += sampleData[pos] * sinc;
            }
        }
        outputTo[toPos] = (float)v;
    }

    /*
    for (int i=0; i<sampleTotalTo; ++i) {
        printf("i=%6d rPos=%6d fraction=%+6.2f output=%f\n",
            i, resamplePosArray[i], fractionArray[i], outputTo[i]);
    }
    printf("resampled\n");
    */

//end:
    delete [] sinPreComputeArray;
    sinPreComputeArray = nullptr;

    delete [] fractionArray;
    fractionArray = nullptr;

    delete [] resamplePosArray;
    resamplePosArray = nullptr;

    return hr;
}
